kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/dates.kt

```diff
@@ -59,12 +60,20 @@ fun Arb.Companion.localDate(
    minDate > maxDate -> throw IllegalArgumentException("minDate must be before or equal to maxDate")
    minDate == maxDate -> Arb.constant(minDate)
    else -> {
-      val leapYears = (minDate.year..maxDate.year).filter { isLeap(it.toLong()) }
-
-      val february28s = leapYears.map { LocalDate.of(it, 2, 28) }
-      val february29s = february28s.map { it.plusDays(1) }
+      val firstPotentialLeapYear = (minDate.year + 3) and (0.inv() shl 2)
+      val potentialLeapYears = when {
+         maxDate.year < firstPotentialLeapYear -> 0
+         else -> (maxDate.year - firstPotentialLeapYear) / 4 + 1
+      }
 
-      arbitrary(february28s + february29s + minDate + maxDate) {
+      arbitrary(edgecaseFn = { rs ->
+         val r = rs.random.nextInt(potentialLeapYears + 1)
+         when {
+            r == 0 -> if (rs.random.nextBoolean()) minDate else maxDate
+            else -> LocalDate.of(firstPotentialLeapYear + (r - 1) * 4, 2, 28)
+               .plusDays(rs.random.nextLong(2))
+         }
+      }) {
```

Comment:
```suggestion
         val  leapYear = rs.random.nextInt(1..potentialLeapYears) * 4
         when {
            r == 0 -> if (rs.random.nextBoolean()) minDate else maxDate
            else -> LocalDate.of(leapYear, 2, 28).plusDays(rs.random.nextLong(0..1))
         }
      }) {
```

Reply:
Fairly sure `firstPotentialLeapYear` got lost in the suggestion, didn't it?
Also the case `r == 0` is gone now. Both the +1 and the -1 were important. (number of potential leap years, two cases each, PLUS the two cases for min and max - even distribution)

Means, the random `leapYear` now obstructs the purpose of both variable and the + 1.
I'm struggling to come up with a concise yet readable alternative... but how about this?

```
val edgeCase = rs.random.nextInt(potentialLeapYears + 1)
when {
  edgeCase == 0 -> if (rs.random.nextBoolean()) minDate else maxDate
  else -> (firstPotentialLeapYear + (edgeCase - 1) * 4).let { potentialLeapYear -> LocalDate.of(potentialLeapYear, 2, 28).plusDays(rs.random.nextLong(0..1))
}
```

or 

```
val edgeCase = rs.random.nextInt(potentialLeapYears + 1)
when {
  edgeCase == 0 -> if (rs.random.nextBoolean()) minDate else maxDate
  else -> LocalDate.of(firstPotentialLeapYear, 2, 28).plusYears((edgeCase - 1) * 4).plusDays(rs.random.nextLong(0..1))
}
```

Reply:
What do you think about adding a new arbitrary-builder which lets us provide edgecases as a `Sequence` instead? Then you could do yield the minDate + maxDate first and then focus on random leap years afterwards?

I'd like to get rid of the +/- 1, since off by one errors are tricky. Is there any way to do that?

Reply:
While I like the idea, having a sequence would basically fix the order of edge cases, and I'm wondering if that's a good or a bad thing. This brings up some more fundamental questions:
- Are all edge-cases equal?
- If not, how to define their weight?
- How could users override their weight?

In this particular case, any single edge-case has equal weight. That's nice if you want to cover as many different edge-case dates as possible. However, if you have a really wide date range, the likelihood of min and max still being generated is getting _really_ small.

So in this particular case, it's maybe best to change the approach, and first classify edge cases and then keep an equal distribution between those. We might even want to add more edge cases. E.g.:
1) actual edges (min/max)
2) (potential) leap years (28, day after 28)
3) first and last days of a month/year/decade/century/millennium (that would be new - but maybe too much?)

However, I'd at least like to go forward and simply split between min/max and leap-year-cases equally. If I can find an easy solution for it, I also want to separate the "exceptional years" so that those are tested more often. Either option also eliminates the +/- 1.

Any thoughts?

Reply:
I think that's a good approach. It feels like one would expect min/max to be tested and not get diluted to the point of not being tested by the presence of leap years. So I think we can consider there to be classes of edge cases, and some more important than others.. I'm not sure if there's any built-in way that emulates this currently. @myuwono do you know?

I think your suggestion of testing min/max and leap-years equally sounds like a good solution in this case. 🚀

Reply:
Just to confirm that I didn't forget about this; unfortunately didn't get around to it yet, but will eventually.